src/Angor/Avalonia/Angor.Contexts.Funding/Projects/Infrastructure/Impl/ProjectInvestmentsRepository.cs

```diff
@@ -0,0 +1,145 @@
+using Angor.Contexts.Funding.Projects.Domain;
+using Angor.Contexts.Funding.Projects.Infrastructure.Interfaces;
+using Angor.Shared;
+using Angor.Shared.Models;
+using Angor.Shared.Services;
+using Blockcore.Consensus.TransactionInfo;
+using CSharpFunctionalExtensions;
+using System.Reactive.Linq;
+using System.Reactive.Threading.Tasks;
+using Angor.Shared.Protocol;
+using Stage = Angor.Shared.Models.Stage;
+
+namespace Angor.Contexts.Funding.Projects.Infrastructure.Impl;
+
+public class ProjectInvestmentsRepository(IProjectRepository projectRepository, INetworkConfiguration networkConfiguration,
+    IIndexerService indexerService, IInvestorTransactionActions investorTransactionActions) : IProjectInvestmentsRepository
+{
+    public async Task<Result<IEnumerable<StageData>>> ScanFullInvestments(string projectId)
+    {
+        var project = await projectRepository.Get(new ProjectId(projectId));
+
+            if (project.IsFailure)
+                return Result.Failure<IEnumerable<StageData>>("Failed to retrieve project data.");
+            
+            var network = networkConfiguration.GetNetwork();
+
+            try
+            {
+                var projectInvestments = await indexerService.GetInvestmentsAsync(projectId);
+                
+                var investments = await projectInvestments
+                    .ToObservable()
+                    .Select(x => indexerService.GetTransactionHexByIdAsync(x.TransactionId))
+                    .Select(hex => network.CreateTransaction(hex.Result))
+                    .Select(trx => indexerService.GetTransactionInfoByIdAsync(trx.GetHash().ToString())
+                        .ToObservable()
+                        .Select(trxInfo => (trx, trxInfo)))
+                    .Merge()
+                    .ToList();
+
+                var stageDataList = await project.Value.Stages
```

Comment:
this is quite a complicated method, if it throws how do you know where it came from?

Reply:
Good point, in the desktop we should not have exceptions the same way, we need to handle them with Result objects, I will add it as a task to handle in here after merging to main.